File: complete/src/main/java/com/example/restservicecors/GreetingController.java
package com.example.restservicecors;

import java.util.concurrent.atomic.AtomicLong;

import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class GreetingController {

	private static final String template = "Hello, %s!";
	private final AtomicLong counter = new AtomicLong();

	@CrossOrigin(origins  = "http://localhost:8080")
	@PostMapping("/greeting")
	public Greeting greeting(@RequestParam(required=false, defaultValue="World") String name) {
		System.out.println("==== in greeting ====");
		return new Greeting(counter.incrementAndGet(), String.format(template, name));
	}

	@PostMapping("/greeting-javaconfig")
	public Greeting greetingWithJavaconfig(@RequestParam(required=false, defaultValue="World") String name) {
		System.out.println("==== in greeting ====");
		return new Greeting(counter.incrementAndGet(), String.format(template, name));
	}

}

File: complete-kotlin/src/main/kotlin/com/example/restservicecors/GreetingController.kt
package com.example.restservicecors

import org.springframework.web.bind.annotation.CrossOrigin
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.util.concurrent.atomic.AtomicLong

@RestController
class GreetingController {

    private val template = "Hello, %s!"
    private val counter = AtomicLong()

    @CrossOrigin(origins = ["http://localhost:9000"])
    @GetMapping("/greeting")
    fun greeting(@RequestParam(required = false, defaultValue = "World") name: String): Greeting {
        println("==== get greeting ====")
        return Greeting(counter.incrementAndGet(), String.format(template, name))
    }

    @GetMapping("/greeting-javaconfig")
    fun greetingWithJavaconfig(@RequestParam(required = false, defaultValue = "World") name: String): Greeting {
        println("==== in greeting ====")
        return Greeting(counter.incrementAndGet(), String.format(template, name))
    }
}
